modules/internal/ChapelDomain.chpl

```diff
@@ -2988,4 +2989,116 @@ module ChapelDomain {
 
   }  // record _domain
 
+  //
+  // Support for creating domains using tuples of bounds
+  //
+
+  /* Creates a rectangular domain with bounds defined by the scalar values `low`
+      and `high`. If `inclusive` is true, the domain includes the `high` value.
+      Otherwise, the domain excludes the `high` value.
+   */
+  proc makeRectangularDomain(low: ?t1, high: ?t2, param inclusive: bool = true)
+    where isValidRangeIdxType(t1) && isValidRangeIdxType(t2)
+  {
+    return if inclusive then {low..high} else {low..<high};
+  }
+
+  @chpldoc.nodoc
+  proc makeRectangularDomain(low: ?t1, high: ?t2, param inclusive: bool = true)
+    where isTuple(t1) && isTuple(t2) &&
+          !(isHomogeneousTuple(low) && isHomogeneousTuple(high))
+  {
+    compilerError("Domains defined using tuple bounds must use homogenous tuples, but got '" +
+                  low.type:string + "' and '" + high.type:string + "'");
+  }
+
+  @chpldoc.nodoc
+  proc makeRectangularDomain(low: ?t1, high: ?t2, param inclusive: bool = true)
+    where isTuple(low) && isTuple(high) &&
+          isHomogeneousTuple(low) && isHomogeneousTuple(high) &&
+          low.size != high.size {
+    compilerError("Domains defined using tuple bounds must use tuples of the same length, " +
+                  "but got '" + low.type:string + "' and '" + high.type:string + "'");
+  }
+
+  @chpldoc.nodoc
+  proc makeRectangularDomain(low: ?t1, high: ?t2, param inclusive: bool = true)
+    where isTuple(low) && isTuple(high) &&
+          isHomogeneousTuple(low) && isHomogeneousTuple(high) &&
+          low.size == high.size &&
+          !(isCoercible(low(0).type, high(0).type) ||
+            isCoercible(high(0).type, low(0).type)) {
+    compilerError("Domains defined using tuple bounds must use tuples of coercible types. " +
+                  "Cannot coerce between '" + low(0).type:string + "' and '" +
+                  high(0).type:string + "'");
+  }
```

Comment:
I think you can reduce some of the duplication here by combining these overloads into one. Fundamentally, the `where` clauses you're defining in each successive overload here are like else-ifs. "Are they not tuples? Else, are they not homogenous? Else, are they not coercible? else, valid overload". It seems encoding it as an actual else-if is cleaner. The steps are:

1. Put the error handling logic into the body of a single overload using actual else-ifs (combining the three functions I've highliighted).
2. Just running step 1 makes the error handling overload a generic function with no where clause, which I think makes it less specific than the function _with_ a where clause, so you may be done. Otherwise -- because it's generic -- it may match valid cases and cause either extraneous errors or cause ambiguity. In that case, you might want to mark it as `pragma "last resort"`.